### PR TITLE
Gateway support via new @apollo/server-gateway-interface package

### PR DESCRIPTION
**Pull request by glasser**
Until now, the refactored AS4 did not support Apollo Gateway (or any
implementation of the AS3 `gateway` option). That's because
`GraphQLRequestContext` is part of the API between Apollo Gateway and
Apollo Server, and that type has changed in some minor but incompatible
ways in AS4.

(Additionally, clashes between `declare module` declarations in AS3 and
AS4 caused issue, but we removed those declarations from this branch in
PRs https://github.com/apollographql/apollo-server/pull/6764 and https://github.com/apollographql/apollo-server/pull/6759.)

This commit restores gateway support. It does this by having AS4 produce
an AS3-style request context object. It uses a new
`@apollo/server-gateway-interface` package to define the appropriate
types for connecting to Gateway.

(Note: this package will be code reviewed in this PR in this repo, but
once it's approved, it will move to the apollo-utils repo and be
released as non-alpha there. Once we've released AS4.0.0 we can move
that package back here, but trying to do some prereleases and some
non-prereleases on the same branch seems challenging.)

This PR removes the top-level `executor` function, which is redundant
with the `gateway` option. (Internally, the relevant field is now named
`gatewayExecutor`.)

Some types had been parametrized by `TContext`, because in AS3,
`GraphQLExecutor` (now `GatewayExecutor`) appeared to take a
`<TContext>`. However, even though the type itself took a generic
argument, its main use in the return from `gateway.load` implicitly
hardcoded the default `TContext`. So we are doubling down on that and
only allowing `GraphQLExecutor` to use AS3's default `TContext`, the
quite flexible `Record<string, any>`.

Most of the way toward https://github.com/apollographql/apollo-server/issues/6719.

### CODE REVIEW

File: docs/source/migration.mdx [hunk 1]
```diff
@@ -741,6 +741,45 @@ new ApolloServer<MyContext>({
 });
 ```
 
+### `executor`
+
+In Apollo Server 3, there are two different ways to specify a replacement for `graphql-js`'s execution functionality. Both of them involve defining a function of the type `GraphQLExecutor`. One way is to specify that function directly as the `executor` constructor option. The other way involves using the `gateway` option.
+
+In Apollo Server 4, this redundancy has been removed: there is no longer an `executor` constructor option. (Additionally, the TypeScript `GraphQLExecutor` type has been renamed `GatewayExector` and moved to the `@apollo/server-gateway-interface` package.)
```
Comment (trevor-scheer) [inline docs/source/migration.mdx:748]:
```suggestion
In Apollo Server 4, this redundancy has been removed: there is no longer an `executor` constructor option. (Additionally, the TypeScript `GraphQLExecutor` type has been renamed `GatewayExecutor` and moved to the `@apollo/server-gateway-interface` package.)
```

File: packages/server-gateway-interface/package.json [hunk 2]
```diff
@@ -0,0 +1,30 @@
+{
+  "name": "@apollo/server-gateway-interface",
+  "version": "0.0.0",
+  "description": "Interface used to connect Apollo Gateway to Apollo Server",
+  "type": "module",
```
Comment (trevor-scheer) [inline packages/server-gateway-interface/package.json:5]: I think we can drop this given no runtime?
```suggestion
```

File: packages/server-gateway-interface/src/index.ts [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+  extensions?: Record<string, any>;
+  http?: GatewayHTTPResponse;
+}
+
+export interface GatewayHTTPResponse {
+  readonly headers: GatewayHTTPHeaders;
+  status?: number;
+}
+
+export interface GatewayHTTPHeaders {
```
Comment (trevor-scheer) [inline packages/server-gateway-interface/src/index.ts:94]: Can this just be a re-export of `FetcherHeaders`?

File: packages/server/src/utils/makeGatewayGraphQLRequestContext.ts [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+// but you can also use `formatError` or `didEncounterErrors` for this perhaps?
+// In any case, AS4 doesn't track `includeStacktraceInErrorResponses` on a
+// per-operation basis; if we find a use case for this we can add it later.
+//
+// So we'll just ignore changes to `logger` and `debug`.
+//
+// Next, there's `request`. We don't know of a use case for mutating the
+// *request* at execution time. If there was a real use case, we could add a
+// function that copies pieces back from the gateway `request` to the AS4
+// request, but we're not bothering to yet. and `response`.
```
Comment (trevor-scheer) [inline packages/server/src/utils/makeGatewayGraphQLRequestContext.ts:73]:
```suggestion
// request, but we're not bothering to yet.
```

File: packages/server-gateway-interface/package.json (hunk 2 above)
Comment (glasser) [reply packages/server-gateway-interface/package.json:5]: Are you sure that TSC doesn't look at this field? I guess that's just a nodenext thing? Do you think it hurts?

Comment (trevor-scheer) [reply packages/server-gateway-interface/package.json:5]: Fair question. I don't think it hurts, and I can't say for sure about `nodenext` so I don't mind if you leave it.

File: packages/server-gateway-interface/src/index.ts (hunk 3 above)
Comment (glasser) [reply packages/server-gateway-interface/src/index.ts:94]: Ooh, nice, I think you're right. FetcherHeaders also has `extends Iterable<[string, string]>` but that shouldn't hurt.

Comment (glasser) [reply packages/server-gateway-interface/src/index.ts:94]: (Request and Response are intentionally smaller though, but you didn't suggest that :) )

**Comment by glasser** [comment]
Trevor said it would be fine to merge (and he's about to be out for a week)

**Pull request merged**